File: algorithm/reverse_linked_list.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# https://leetcode.com/problems/reverse-linked-list/


class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def reverse_list(self, head):
        tail = None
        while head:
            new_head = head
            head = head.next
            new_head.next = tail
            tail = new_head
        return tail

if __name__ == '__main__':
    head = ListNode(1)
    head1 = ListNode(2)
    head2 = ListNode(3)
    head3 = ListNode(4)
    head.next = head1
    head1.next = head2
    head2.next = head3
    new_list = Solution().reverse_list(head)
    while new_list:
        print new_list.val
        new_list = new_list.next

File: algorithm/reverse_linked_list.c
//
// https://leetcode.com/problems/reverse-linked-list/
//

#include <stdio.h>
#include <stdlib.h>

struct ListNode {
    int val;
    struct ListNode * next;
};

struct ListNode* reverseList(struct ListNode * head);

int main()
{
    struct ListNode *head = (struct ListNode*)malloc(sizeof(struct ListNode));
    struct ListNode *head1 = (struct ListNode*)malloc(sizeof(struct ListNode));
    struct ListNode *head2 = (struct ListNode*)malloc(sizeof(struct ListNode));
    struct ListNode *new_head = (struct ListNode*)malloc(sizeof(struct ListNode));
    head->val = 1;
    head->next = head1;
    head1->val = 2;
    head1->next = head2;
    head2->val = 3;
    head2->next = NULL;
    new_head = reverseList(head);
    while(new_head){
        printf("%d ", new_head->val);
        new_head = new_head->next;
    }
    return 0;
}

struct ListNode* reverseList(struct ListNode * head)
{
    struct ListNode* tail = NULL;
    struct ListNode* new_head = NULL;
    while(head){
        new_head = head;
        head = head->next;
        new_head->next = tail;
        tail = new_head;
    }
    tail = NULL;
    return new_head;
}
